### ISSUE TITLE
Towards a more modular odoc

### ISSUE THREAD

**Issue by Drup** [ocaml/odoc#431]
Now that we can have several output, it's time to think on how to have outputs living outside of odoc, which means we must make odoc more modular. There are several existing tickets on the question, but I don't think anything concrete got out of those and I think it deserves some centralized discussion. 

cc @rgrinberg and @jeremiedimino as I expect them to have opinion on the tooling aspects.

I see several path we could take:
- Make a plugin system, à la ocamldoc. This means making the code part from IR to rendering independent from the model, expose all that as a library, and make some dynlink magic.
  **pros**:
  - Using the plugin is usually rather easy (probably just a cli argument in odoc).

  **cons**:
  - Plugins systems can be brittle and building plugins is often a mess.
  - Does not immediately provide caching of the IR

- Use a standard interchange format (probably the IR itself, marshalled) and decouple the pipeline in CM* -> IR and IR -> OUTPUT. IIRC, this is how the ancestor codoc worked.
  **pros**:
  - Nice immediate caching!
  - Easy to build outputs
 
  **cons**:
  - Make the pipeline more complicated.
  - Make the current pipeline slower, as we need to serialize more.

- Don't make things modular at all. Ask everyone to rebuild a mini-odoc with their pet output and redo the work from scratch every time.
  **cons**
  - We still need to expose most of odoc's guts as a library, so it's not really any simpler.
  - No caching

We could also decide to merge all the output into odoc, but I don't think that's any realistic long-term.

Opinions ? Alternative ideas ?

**Comment by jonludlam** [comment]
The current plan is effectively number 2 - we output the IR (`Odoc_model.Lang.Compilation_unit.t`) after the link phase (See https://github.com/ocaml/odoc/wiki/Model-design#compiling-and-linking-phases) and have the html generation phase take that as input. The disadvantage of this, as far as a modular system goes, is, of course, that the on-disk format is very brittle. We could opt for something a bit more interoperable if we're truly going to decouple the generation of different outputs from the main odoc binary. It's not inconceivable that we could use json for this, and that would address the reason-association project too - https://www.reason-association.org/projects/doc-tools

**Comment by aantron** [comment]
cc @rizo @ryyppy

**Comment by Drup** [comment]
> The current plan is effectively number 2 - we output the IR (`Odoc_model.Lang.Compilation_unit.t`) after the link phase (See https://github.com/ocaml/odoc/wiki/Model-design#compiling-and-linking-phases) and have the html generation phase take that as input.

Wait, is that already the case ? I was convinced otherwise .... In that case, that makes it easy: we just switch to using the lower IR (starting now, I will say "the document representation", as opposed to "the model").

**Comment by jeremiedimino** [comment]
I'll let you discuss the odoc design, but just a note regarding a plugin system: if you go in that direction, personally i would go for a "static" system where the build system links static custom odoc executables with all the plugins statically linked in. It requires a bit more build system collaboration, but this kind of things already works pretty well in Dune. It's also more portable.

**Comment by jonludlam** [comment]
Not _yet_ the case (not even in my `next` branch), but that's the way the code is structured. We'll probably want to adjust the types (e.g. the paths/fragments/references etc are all a bit hard to interpret) but that's probably not too tricky.

**Comment by lpw25** [comment]
> In that case, that makes it easy: we just switch to using the lower IR (starting now, I will say "the document representation", as opposed to "the model").

I think that it is probably better to keep saving the model to a file rather than the document representation. It is needed to produce outputs that are not documents. For example, things like search indices probably want access to the model itself.

I think that we should consider the question of what gets saved to disk and what is the public API as separate questions. It is perfectly reasonable to save the model to disk, provide users with the "model -> document representation" function in a library and treat the document representation as the public API. I would think of the decision of what to save as more about maximizing incrementality and sharing than which interface is public.

> We could also decide to merge all the output into odoc, but I don't think that's any realistic long-term.

I think this option should be given serious consideration. In practice the number of interesting output formats is not that high, and the cost of maintaining such outputs is probably low. The benefit of not needing to maintain any form external compatibility, especially whilst the system is still evolving, should not be underestimated.

**Comment by Drup** [comment]
> In practice the number of interesting output formats is not that high

That is self-fulfilling: If we don't make outputs easy to create out of odoc, there will not be many outputs. IIUC the reason/BS people already use 2+ (docausorus and standalone fancy non-odoc HTML) and they wanted to have a json (not for search) one. I might want to have a wiki creole output for eliom, etc.

> the cost of maintaining such outputs is probably low.

Except for the high-added-value one that they have in the reason ecosystem. Especially since practice has shown that if people have to merge they work, they'll just fork/redevelop things instead.

> The benefit of not needing to maintain any form external compatibility

Once other people have thoroughly bikeshed the document representation and the dust settles down, I doubt the documentation representation is going to move much and we could easily version it if that starts becoming an issue.

**Comment by Drup** [comment]
I think with the various cleanup of the API, we are getting pretty close to being able to build a standalone "renderer" that could execute the step "model -> output". After that, it's just a matter of formalizing all the tools to be able to call such standalone renderer instead of the built-in calls.

**Comment by Julow** [comment]
Now that the pipeline is splitted into 3 steps (compile, link and generate), concerns about caching and driving are gone. Odoc no longer worry about compiling, linking and calling the right generator, that's the driver's job (eg. dune, odig).

I'd propose an iteration of the idea 3:
- Generators link against `odoc.document` and a small new library that calls the right internal functions (loading of `.odocl` files, call the user's `Odoc_document.GENERATOR`).
- Generators are binaries with the same API as `odoc html-generate -o <output dir> <input.odocl>` and it's not Odoc's concern to hook them into the build system. 

That way, generators still need to link against Odoc's libraries but won't have to deal with the model (which API isn't stable).

**Comment by Drup** [comment]
Yes, that's close to what I had in mind and seems the best solution to me overall.

**Comment by jonludlam** [comment]
This is pretty similar to what's happening in the generator in [voodoo](https://github.com/jonludlam/voodoo/tree/main/gen)

**Comment by zoggy** [comment]
Regarding "plugins" or whatever you call it, an important point IMO is that they should be composable: if two "plugins" bring some cool features for a given output, then they should be usable together to get all these cool features in this output (of course if the features they provide do not interfere with each other).
This was achieved in ocamldoc using classes in first-class modules. I did not use odoc yet and I did not look at the code, so I don't know if that is easily possible.
